### PR TITLE
tpu quic integration test on main

### PR DESCRIPTION
**Pull request by grooviegermanikus**
this adds a basic integration test for lite-rpc TPU client code against the solana quic streamer.
the test can serve as basis for testing the connection handling as well as performance testing.

### CODE REVIEW

File: services/tests/literpc_tpu_quic_server_integrationtest.rs [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+        .worker_threads(16) // note: this value has changed with the "deadlock fix" - TODO experiment with it
+        .enable_all()
+        .build()
+        .expect("failed to build tokio runtime for lite-rpc-tpu-client");
+
+    let literpc_validator_identity = Arc::new(Keypair::new());
+    let udp_listen_socket = UdpSocket::bind("127.0.0.1:0").unwrap();
+    let listen_addr = udp_listen_socket.local_addr().unwrap();
+
+    let (inbound_packets_sender, inbound_packets_receiver) = crossbeam_channel::unbounded();
```
Comment (godmodegalactus) [inline services/tests/literpc_tpu_quic_server_integrationtest.rs:73]: Cross beam usually have many issues and does not run well with tokio. Make it mpsc::UnboundSender.

File: services/tests/literpc_tpu_quic_server_integrationtest.rs [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+    // TODO this is a race
+    sleep(Duration::from_millis(1500)).await;
+
+    for i in 0..SAMPLE_TX_COUNT {
+        let raw_sample_tx = build_raw_sample_tx(i);
+        debug!(
+            "broadcast transaction {} to {} receivers: {}",
+            raw_sample_tx.0,
+            broadcast_sender.receiver_count(),
+            format!("hi {}", i)
```
Comment (godmodegalactus) [inline services/tests/literpc_tpu_quic_server_integrationtest.rs:306]: format message correctly.

File: services/tests/literpc_tpu_quic_server_integrationtest.rs [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+            for packet in packet_batch.iter() {
+                let tx = packet
+                    .deserialize_slice::<VersionedTransaction, _>(..)
+                    .unwrap();
+                trace!(
+                    "read transaction from quic streaming server: {:?}",
+                    tx.get_signature()
+                );
+                count_map.insert_or_increment(*tx.get_signature());
+                // for ix in tx.message.instructions() {
```
Comment (godmodegalactus) [inline services/tests/literpc_tpu_quic_server_integrationtest.rs:157]: remove all commented code.

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

**Comment by godmodegalactus** [comment]
Did you do clippy and fmt?

File: services/tests/literpc_tpu_quic_server_integrationtest.rs [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+        .worker_threads(1)
+        .thread_name("quic-server")
+        .enable_all()
+        .build()
+        .expect("failed to build tokio runtime for testing quic server");
+
+    // lite-rpc
+    let runtime_literpc = tokio::runtime::Builder::new_multi_thread()
+        // see lite-rpc -> main.rs
+        .worker_threads(16) // note: this value has changed with the "deadlock fix" - TODO experiment with it
```
[the remark this answers is not in the document]
Comment (grooviegermanikus) [reply services/tests/literpc_tpu_quic_server_integrationtest.rs:64]: what exactly was the deadlock fix ? do you remember - will try with 1 in the meantime

[the remark this answers is not in the document]
Comment (grooviegermanikus) [reply services/tests/literpc_tpu_quic_server_integrationtest.rs:64]: ok .. with "1" it works

File: services/tests/literpc_tpu_quic_server_integrationtest.rs (hunk 1 above)
Comment (grooviegermanikus) [reply services/tests/literpc_tpu_quic_server_integrationtest.rs:73]: ma :) solana-quic-streamer brings in that dependency

File: services/tests/literpc_tpu_quic_server_integrationtest.rs (hunk 2 above)
Comment (grooviegermanikus) [reply services/tests/literpc_tpu_quic_server_integrationtest.rs:306]: will remove the debug. it's quite worthless and confusing

File: services/tests/literpc_tpu_quic_server_integrationtest.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (godmodegalactus) [reply services/tests/literpc_tpu_quic_server_integrationtest.rs:64]: Yeah actually with 1 whenever we are not correctly scheduling stuff it deadlocks.

File: services/tests/literpc_tpu_quic_server_integrationtest.rs (hunk 1 above)
Comment (godmodegalactus) [reply services/tests/literpc_tpu_quic_server_integrationtest.rs:73]: Ahh yes sad.

File: services/tests/literpc_tpu_quic_server_integrationtest.rs (hunk 3 above)
Comment (grooviegermanikus) [reply services/tests/literpc_tpu_quic_server_integrationtest.rs:157]: done

File: services/tests/literpc_tpu_quic_server_integrationtest.rs [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+use tokio::runtime::{Builder, Runtime};
+use tokio::sync::broadcast;
+use tokio::sync::broadcast::error::SendError;
+use tokio::task::JoinHandle;
+use tokio::time::{interval, sleep};
+use tokio::{join, spawn, task};
+use tracing_subscriber::{filter::LevelFilter, fmt};
+
+// note: logging will be auto-adjusted
+const SAMPLE_TX_COUNT: u32 = 20;
```
[the remark this answers is not in the document]
Comment (grooviegermanikus) [reply services/tests/literpc_tpu_quic_server_integrationtest.rs]: done

File: services/Cargo.toml [hunk 6]
```diff
@@ -35,4 +35,12 @@ async-channel = { workspace = true }
 quinn = { workspace = true }
 chrono = { workspace = true }
 rustls = { workspace = true }
-solana-lite-rpc-core = { workspace = true }
\ No newline at end of file
+solana-lite-rpc-core = { workspace = true }
+tracing-subscriber = { workspace = true, features = ["std", "env-filter"] }
```
Comment (godmodegalactus) [inline services/Cargo.toml:39]: why did not you move this to dev-dependencies?

File: services/tests/literpc_tpu_quic_server_integrationtest.rs [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+#[derive(Copy, Clone, Debug)]
+struct TestCaseParams {
+    sample_tx_count: u32,
+    stake_connection: bool,
+}
+
+const MAXIMUM_TRANSACTIONS_IN_QUEUE: usize = 200_000;
+const MAX_QUIC_CONNECTIONS_PER_PEER: usize = 8; // like solana repo
+
+const QUIC_CONNECTION_PARAMS: QuicConnectionParameters = QuicConnectionParameters {
```
Comment (godmodegalactus) [inline services/tests/literpc_tpu_quic_server_integrationtest.rs:51]: Should be exactly like lite-rpc, could you make const variable for both of them, usually most of them are already constant.

File: services/tests/literpc_tpu_quic_server_integrationtest.rs [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+pub fn many_transactions() {
+    configure_logging(false);
+
+    wireup_and_send_txs_via_channel(TestCaseParams {
+        sample_tx_count: 10000,
+        stake_connection: true,
+    });
+}
+
+#[ignore]
```
Comment (godmodegalactus) [inline services/tests/literpc_tpu_quic_server_integrationtest.rs:91]: why ignore?

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

**Comment by godmodegalactus** [comment]
Can still see lot of warning when I compile the code

```
warning: unused import: `anyhow::bail`
 --> services/tests/literpc_tpu_quic_server_integrationtest.rs:1:5
  |
1 | use anyhow::bail;
  |     ^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused imports: `Receiver`, `RecvError`, `RecvTimeoutError`
 --> services/tests/literpc_tpu_quic_server_integrationtest.rs:3:25
  |
3 | use crossbeam_channel::{Receiver, RecvError, RecvTimeoutError, Sender};
  |                         ^^^^^^^^  ^^^^^^^^^  ^^^^^^^^^^^^^^^^

warning: unused import: `futures::future::join_all`
 --> services/tests/literpc_tpu_quic_server_integrationtest.rs:4:5
  |
4 | use futures::future::join_all;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `error`
 --> services/tests/literpc_tpu_quic_server_integrationtest.rs:5:18
  |
5 | use log::{debug, error, info, trace, warn};
  |                  ^^^^^

warning: unused import: `quinn::TokioRuntime`
 --> services/tests/literpc_tpu_quic_server_integrationtest.rs:6:5
  |
6 | use quinn::TokioRuntime;
  |     ^^^^^^^^^^^^^^^^^^^

warning: unused import: `solana_sdk::signer::keypair`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:17:5
   |
17 | use solana_sdk::signer::keypair;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `std::ops::Deref`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:26:5
   |
26 | use std::ops::Deref;
   |     ^^^^^^^^^^^^^^^

warning: unused import: `std::option::Option`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:27:5
   |
27 | use std::option::Option;
   |     ^^^^^^^^^^^^^^^^^^^

warning: unused import: `std::path::Path`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:28:5
   |
28 | use std::path::Path;
   |     ^^^^^^^^^^^^^^^

warning: unused import: `AtomicU32`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:30:37
   |
30 | use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
   |                                     ^^^^^^^^^

warning: unused import: `Runtime`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:34:31
   |
34 | use tokio::runtime::{Builder, Runtime};
   |                               ^^^^^^^

warning: unused import: `tokio::sync::broadcast`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:35:5
   |
35 | use tokio::sync::broadcast;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `tokio::sync::broadcast::error::SendError`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:36:5
   |
36 | use tokio::sync::broadcast::error::SendError;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `interval`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:38:19
   |
38 | use tokio::time::{interval, sleep};
   |                   ^^^^^^^^

warning: unused imports: `filter::LevelFilter`, `fmt`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:40:26
   |
40 | use tracing_subscriber::{filter::LevelFilter, fmt};
   |                          ^^^^^^^^^^^^^^^^^^^  ^^^

warning: unused doc comment
   --> services/tests/literpc_tpu_quic_server_integrationtest.rs:128:9
    |
128 |           /// setup solana Quic streamer
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
129 |           // see log "Start quic server on UdpSocket { addr: 127.0.0.1:xxxxx, fd: 10 }"
130 | /         let staked_nodes = StakedNodes {
131 | |             total_stake: 100,
132 | |             max_stake: 40,
133 | |             min_stake: 0,
...   |
141 | |             },
142 | |         };
    | |__________- rustdoc does not generate documentation for statements
    |
    = help: use `//` for a plain comment
    = note: `#[warn(unused_doc_comments)]` on by default

warning: unused import: `tracing_subscriber::util::SubscriberInitExt`
  --> services/tests/literpc_tpu_quic_server_integrationtest.rs:39:5
   |
39 | use tracing_subscriber::util::SubscriberInitExt;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused variable: `tpu_addr`
   --> services/tests/literpc_tpu_quic_server_integrationtest.rs:381:9
    |
381 |     let tpu_addr = SocketAddr::new(addr, port);
    |         ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_tpu_addr`
    |
    = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `tpu_addr`
   --> services/tests/literpc_tpu_quic_server_integrationtest.rs:442:13
    |
442 |         let tpu_addr = SocketAddr::new(addr, port);
    |             ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_tpu_addr`

warning: constant `NUM_QUIC_STREAMER_WORKER_THREADS` is never used
   --> services/tests/literpc_tpu_quic_server_integrationtest.rs:105:11
    |
105 |     const NUM_QUIC_STREAMER_WORKER_THREADS: usize = 1;
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: fields `sock`, `exit`, `join_handler`, and `stats` are never read
   --> services/tests/literpc_tpu_quic_server_integrationtest.rs:392:5
    |
391 | struct SolanaQuicStreamer {
    |        ------------------ fields in this struct
392 |     sock: UdpSocket,
    |     ^^^^
393 |     exit: Arc<AtomicBool>,
    |     ^^^^
394 |     join_handler: JoinHandle<()>,
    |     ^^^^^^^^^^^^
395 |     stats: Arc<StreamStats>,
    |     ^^^^^

warning: method `get_socket_addr` is never used
   --> services/tests/literpc_tpu_quic_server_integrationtest.rs:399:12
    |
398 | impl SolanaQuicStreamer {
    | ----------------------- method in this implementation
399 |     pub fn get_socket_addr(&self) -> SocketAddr {
    |            ^^^^^^^^^^^^^^^

warning: method `shutdown` is never used
   --> services/tests/literpc_tpu_quic_server_integrationtest.rs:405:18
    |
404 | impl SolanaQuicStreamer {
    | ----------------------- method in this implementation
405 |     pub async fn shutdown(self) {
    |                  ^^^^^^^^

warning: `solana-lite-rpc-services` (test "literpc_tpu_quic_server_integrationtest") generated 23 warnings (run `cargo fix --test "literpc_tpu_quic_server_integrationtest"` to apply 17 suggestions)
```

**Pull request merged**